### PR TITLE
Update to the recent version of quda

### PR DESCRIPTION
**Pull request by sy3394**

### CODE REVIEW

File: lyncs_quda/array.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        return self._qarray
+
+    def __len__(self):
+        return self._size
+
+    def __getitem__(self, i):
+        # TODO: suport slicing
+        return self.qarray[i]
+
+    def __setitem(self, i, val):
```
Comment (sbacchio) [inline lyncs_quda/array.py]: change to `__setitem__`

File: lyncs_quda/array.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        self._qarray = lib.quda.array[typename, size]()
+
+        if elems != None:
+            if isiterable(elems):
+                if len(elems) > size:
+                    raise ValueError()
+                for i, e in enumerate(elems):
+                    self._qarray[i] = e
+            else:
+                self._qarray[0] = elems
```
Comment (sbacchio) [inline lyncs_quda/array.py]: if only one given maybe we want to set all to that value?

File: lyncs_quda/include/.gitignore [hunk 3]
```diff
@@ -1,3 +0,0 @@
-*
-!.gitignore
-!evenodd.h
```
Comment (sbacchio) [inline lyncs_quda/include/.gitignore]: This should be restored

**Review by sbacchio — CHANGES_REQUESTED** [review]

File: lyncs_quda/array.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, typename, size, elems=None):
+        self._size = size
+        self._qarray = lib.quda.array[typename, size]()
+
+        if elems != None:
+            if isiterable(elems):
+                if len(elems) > size:
+                    raise ValueError()
+                for i, e in enumerate(elems):
+                    self._qarray[i] = e
```
Comment (sbacchio) [inline lyncs_quda/array.py]: Since you define `__setitem__`, then you can use directly `self[i]=e`

File: lyncs_quda/array.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    "Array",
+]
+
+from lyncs_utils import isiterable
+from .lib import lib
+
+
+def lat_dims(elems=[1, 1, 1, 1]):
+    # QUDA_MAX_DIM = 6 by default
+    return Array(int, 6, elems).qarray
```
Comment (sbacchio) [inline lyncs_quda/array.py]: instead of `6` I guess you could use lib.QUDA_MAX_DIM

File: lyncs_quda/array.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                    self[idx] = val
+            else:
+                self[0] = elems
+
+    @property
+    def qarray(self):
+        return self._qarray
+
+    def __len__(self):
+        return self._size
```
Comment (sbacchio) [inline lyncs_quda/array.py]: Would be nice if size is taken directly from qarray

**Comment by sbacchio** [comment]
@sy3394 I took care some of the changes and I left comments that still need to be addressed.
I am going to test it. If everything works fine we can merge it!

**Comment by sbacchio** [comment]
@sy3394 I did some more work on this PR, please have a look.
For now it is working fine with my software stack so we could merge it in soon.

File: lyncs_quda/gauge_field.py [hunk 7]
```diff
@@ -632,21 +627,24 @@ def _check_paths(self, paths):
...
-        for i, path in enumerate(paths):
-            for j, step in enumerate(path):
-                paths_array[0, i, j] = convert(step)
-
-        return paths_array, lengths
+    def _paths_for_sum(self, paths, coeffs):
+        aux = ([], [], [], [])
+        for i, (path, coeff) in enumerate(zip(paths, coeffs)):
+            if path[0] < 0:
+                raise ValueError(f"Path {i} = {path} nevative first movement")
```
Comment (sy3394) [inline lyncs_quda/gauge_field.py]: nevative -> negative

File: lyncs_quda/gauge_field.py [hunk 8]
```diff
@@ -661,16 +669,35 @@ def _paths_for_force(self, paths, coeffs):
...
+
+        lengths = numpy.array(list(map(len, paths[0])), dtype="int32")
+        max_length = max(lengths.max(), 1)
+        ndims = len(paths)
+        npaths = len(paths[0])
+        paths_array = numpy.empty((ndims, npaths, max_length), dtype="int32")
+
+        convert = lambda step: (step - 1) if step > 0 else (8 + step)
+
+        count = defaultdict(int)
```
Comment (sy3394) [inline lyncs_quda/gauge_field.py]: where is this used?

File: lyncs_quda/gauge_field.py [hunk 9]
```diff
@@ -661,16 +669,35 @@ def _paths_for_force(self, paths, coeffs):
...
         force=False,
-        grad=None,
-        left_grad=False,
-        keep_paths=False,
+        sum_paths=True,  # If False, returns a list of gauge fields for every path
+        insertion=None,
+        left=True,
     ):
         """
         Computes the gauge paths on the lattice.
```
Comment (sy3394) [inline lyncs_quda/gauge_field.py:702]: This is a comment for description following this line, which is not showing up here.
It says:
"
- The paths are given with respect to direction "1" and
          this must be the first number of every path list.
        - Directions go from 1 to self.ndims
        - Negative value (-1,...) means backward movement in the direction
        - Paths are then rotated for every direction.
"
Does this suggest that the input "paths" is, e.g., a list of list of integers where the the first element of each list in paths should always be 1?  Then, rotation over other directions are performed on each path.  Is my understanding correct?  If so, this should be checked by _check_paths(self, paths), although the code suggests this is not what is done.

Comment (sbacchio) [reply lyncs_quda/gauge_field.py:702]: Your understanding is correct, but that comment it is outdated. This was the old implementation. I will fix it.. and thanks also for catching the other issues

**Review by sbacchio — APPROVED** [review]
Looks good to me. @sy3394 if you don't have further comments, please merge!

File: lyncs_quda/gauge_field.py [hunk 10]
```diff
@@ -661,101 +669,171 @@ def _paths_for_force(self, paths, coeffs):
...
 
-        The same paths are computed for every direction.
-
-        - The paths are given with respect to direction "1" and
-          this must be the first number of every path list.
-        - Directions go from 1 to self.ndims
-        - Negative value (-1,...) means backward movement in the direction
-        - Paths are then rotated for every direction.
+        The paths are list of integers with the following structure
```
Comment (sy3394) [inline lyncs_quda/gauge_field.py]: How about "A path is a list of ...".  To me, that's easier to grasp.

**Review by sy3394 — COMMENTED** [review]
just one small suggestion for readability.

**Pull request merged**